toolkit/tools/imagegen/installutils/overlay.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package installutils
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"microsoft.com/pkggen/imagegen/diskutils"
+	"microsoft.com/pkggen/internal/logger"
+)
+
+// Overlay Struct representing an overlay mount
+type Overlay struct {
+	DevicePath string
+	lowerDir   string
+	upperDir   string
+	workDir    string
+}
+
+func (o Overlay) getMountArgs() string {
+
+	return fmt.Sprintf("lowerdir=%s,upperdir=%s,workdir=%s", o.lowerDir, o.upperDir, o.workDir)
+}
+
+func (o Overlay) setupFolders() (err error) {
+	err = os.MkdirAll(o.lowerDir, os.ModePerm)
+	if err != nil {
+		logger.Log.Errorf("Could not create directory (%s)", o.lowerDir)
+		return
+	}
+	err = os.MkdirAll(o.upperDir, os.ModePerm)
+	if err != nil {
+		logger.Log.Errorf("Could not create directory (%s)", o.upperDir)
+		return
+	}
+	err = os.MkdirAll(o.workDir, os.ModePerm)
+	if err != nil {
+		logger.Log.Errorf("Could not create directory (%s)", o.workDir)
+		return
+	}
+	err = mount(o.lowerDir, o.DevicePath, "", "")
+	if err != nil {
+		logger.Log.Errorf("Could not mount %s to %s", o.DevicePath, o.lowerDir)
+	}
+	return
+}
+
+func (o Overlay) getUpperDir() (upperDir string) {
+	return o.upperDir
+}
+
+func (o Overlay) unmount() (err error) {
+	err = umount(o.lowerDir)
+	if err != nil {
+		logger.Log.Warnf("Unmount of loopback(%s) failed. Still continuing", o.lowerDir)
+	}
+
+	err = diskutils.DetachLoopbackDevice(o.DevicePath)
+	if err != nil {
+		logger.Log.Warnf("Losetup of loopback(%s) failed. Still continuing", o.lowerDir)
```

Comment:
Is it possible for the above `umount(o.lowerDir)` to fail but not this call? If so the unmount error may get overidden and this routine would return success.

Reply:
Fixed.

---
In reply to: [577070966](https://github.com/microsoft/CBL-Mariner/pull/611#discussion_r577070966) [](ancestors = 577070966)